File: Assets/Code/Logic/VIP/Model/VIPData.cs
using System.Collections.Generic;
using Common.Util;
using Logic.Game.Model;
using Common.Localization;

namespace Logic.VIP.Model
{
    public class VIPData
    {
        private static Dictionary<int, VIPData> _vipDataDictionary;

        public static Dictionary<int, VIPData> VIPDataDictionary
        {
            get
            {
                if (_vipDataDictionary == null)
                {
                    _vipDataDictionary = CSVUtil.Parse<int, VIPData>("config/csv/vip_config", "id");
                }
                return _vipDataDictionary;
            }
        }

        public static VIPData GetVIPData(int vipLevel)
        {
            VIPData vipData = null;
            VIPDataDictionary.TryGetValue(vipLevel, out vipData);
            return vipData;
        }

        public VIPData GetNextLevelVIPData()
        {
            VIPData nextLevelVIPData = null;
            VIPDataDictionary.TryGetValue(id + 1, out nextLevelVIPData);
            return nextLevelVIPData;
        }

        public static List<VIPData> GetAllVIPDataList()
        {
            return VIPDataDictionary.GetValues();
        }

        public bool IsMinLevelVIPData()
        {
            return this.id == VIPDataDictionary.First().Key;
        }

        public bool IsMaxLevelVIPData()
        {
            return this.id == VIPDataDictionary.Last().Key;
        }

        public string ShortName
        {
            get
            {
                return string.Format(Localization.Get("common.vip_short_name"), id);
            }
        }


        //		public static VIPData DoubleSpeedFightOpenVIPData
        //		{
        //			get
        //			{
        //				VIPData vipData = null;
        //				List<VIPData> vipDataList = GetAllVIPDataList();
        //				for (int i = 0, vipDataCount = vipDataList.Count; i < vipDataCount; i++)
        //				{
        //					vipData = vipDataList[i];
        //					if (vipData.canFightInDoubleSpeed)
        //					{
        //						break;
        //					}
        //				}
        //				return vipData;
        //			}
        //		}

        /// <summary>
        /// VIP ID.
        /// </summary>
        [CSVElement("id")]
        public int id;

        /// <summary>
        /// VIP显示名称.
        /// </summary>
        [CSVElement("name")]
        public string name;

        /// <summary>
        /// VIP权限描述.
        /// </summary>
        [CSVElement("des")]
        public string des;

        /// <summary>
        /// VIP福利物品列表.
        /// </summary>
        public List<GameResData> benefitItemList;
        [CSVElement("rewards")]
        public string rewards
        {
            set
            {
                benefitItemList = GameResData.ParseGameResDataList(value);
            }
        }
        public GameResData dailyWelfareSweep;
        [CSVElement("daily_welfare")]
        public string daily_welfare
        {
            set
            {
                dailyWelfareSweep = new GameResData(value);
            }
        }
        /// <summary>
        /// 相应VIP等级所需经验.
        /// </summary>
        [CSVElement("exp")]
        public int exp;

        /// <summary>
        /// 购买远征次数.
        /// </summary>
        [CSVElement("faraway")]
        public int BuyExpeditionTimes;

        /// <summary>
        /// 体力上限点数.
        /// </summary>
        [CSVElement("pve_action_add")]
        public int pveActionAdd;

        /// <summary>
        /// 每日副本购买次数.
        /// </summary>
        [CSVElement("daily_dungeon_buytimes")]
        public int dailyDungeonBuyTimes;

        /// <summary>
        /// PVE行动力购买次数.
        /// </summary>
        [CSVElement("pve_action_buytimes")]
        public int pveActionBuyTimes;

        /// <summary>
        /// 金币购买次数.
        /// </summary>
        [CSVElement("gold_buytimes")]
        public int goldBuyTimes;

        /// <summary>
        /// 竞技场刷新次数.
        /// </summary>
        [CSVElement("pvp_refresh")]
        public int pvpRefreshTimes;

        /// <summary>
        /// 竞技场每日可打次数.
        /// </summary>
        [CSVElement("pvp_add")]
        public int pvpAdd;

        /// <summary>
        /// 世界树行动力上限.
        /// </summary>
        [CSVElement("world_tree_add")]
        public int worldTreeActionAdd;

        /// <summary>
        /// 世界树行动力购买次数.
        /// </summary>
        [CSVElement("world_tree_buytimes")]
        public int worldTreeActionBuyTimes;

        /// <summary>
        /// 普通副本额外金币收益.
        /// </summary>
        [CSVElement("dungeon_gold_add")]
        public int dungeonExtraGoldAmount;

        /// <summary>
        /// 每日副本额外翻牌次数.
        /// </summary>
        [CSVElement("daily_dungeon_add")]
        public int dailyDungeonExtraDrawCardTimes;

        /// <summary>
        /// 阵型培养点购买次数.
        /// </summary>
        [CSVElement("formation_add")]
        public int formation_add;

        /// <summary>
        /// 三倍速.
        /// </summary>
        [CSVElement("three_speed")]
        public bool three_speed;

		[CSVElement("dayRefreshTimes")]
		public int dayRefreshTimes;
    }
}

File: Assets/ToLua/Lua/user/ui/vip/model/vip_data.lua
local t = {}
t.data = {}
t.minLevel = 1
t.maxLevel = 1
local item = {}
item.__index = item

function item.New(table)
	local o = {}
	setmetatable(o,item)
  
	o.index = tonumber(table.index)
	o.id = tonumber(table.id)
	o.name = table.name
	o.des = table.des
	o.rewards = table.rewards
	o.daily_welfare = table.daily_welfare
	o.exp = tonumber(table.exp)
	o.faraway = tonumber(table.faraway)
	o.pve_action_add = tonumber(table.pve_action_add)
	o.daily_dungeon_buytimes = tonumber(table.daily_dungeon_buytimes)
	o.pve_action_buytimes = tonumber(table.pve_action_buytimes)
	o.gold_buytimes = tonumber(table.gold_buytimes)
	o.pvp_refresh = tonumber(table.pvp_refresh)
	o.pvp_add = tonumber(table.pvp_add)
	o.world_tree_add = tonumber(table.world_tree_add)
	o.world_tree_buytimes = tonumber(table.world_tree_buytimes)
	o.dungeon_gold_add = tonumber(table.dungeon_gold_add)
	o.daily_dungeon_add = tonumber(table.daily_dungeon_add)
	o.formation_add = tonumber(table.formation_add)
	o.explore_TaskNum = tonumber(table.explore_TaskNum)
  o.plunder_occ = tonumber(table.plunder_occ)
	o.plunder_num = tonumber(table.plunder_num)
  o.three_speed = table.three_speed
	o.blackmarket_BuyNum = table.blackmarket_BuyNum
	o.dayRefreshTimes = table.dayRefreshTimes
	o.dayGoldHandTimes = tonumber(table.dayGoldHandTimes)
  o.fund_lv_limit = tonumber(table.fund_lv_limit)
	return o
end

function t.GetAllVIPDataList ()
  return t.data
end

function t.GetVIPData (vipLevel)
  return t.data[vipLevel]
end

function t.SortData (aVIPData, bVIPData)
  return aVIPData.id < bVIPData.id
end

function item:GetBenefitItems ()
  local gameResData = require('ui/game/model/game_res_data')
  local benefitItems = gameResData.ParseGameResDataList(self.rewards)
  return benefitItems
end

function item:GetNextLevelVIPData ()
  return t.data[self.id + 1]
end

function item:IsMinLevel ()
  return self.id == t.minLevel
end

function item:IsMaxLevel ()
  return self.id == t.maxLevel
end

function item:IsCanBuyFoundLevelGift()
  return self.fund_lv_limit == 1
end

function item:ShortName ()
  return string.format(LocalizationController.instance:Get('common.vip_short_name_lua'), self.id)
end

local function Start()
	local origin = dofile('vip_config')
	origin.ForEach(function(id,table)
		local newItem = item.New(table)
		t.data[newItem.id] = newItem
	end)

  table.sort(t.data, t.SortData)
  
  for k, v in pairs(t.data) do
    if v.id < t.minLevel then
      t.minLevel = v.id
    end
    if v.id > t.maxLevel then
      t.maxLevel = v.id
    end
  end
end

Start()
return t